### PR TITLE
API: cancel batch change that is pending review

### PR DESCRIPTION
**Pull request by britneywright**
Changes in this pull request:
- Give creators of a batch change the ability to cancel the change if it's approval status is Pending Review

### CODE REVIEW

File: modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala [hunk 1]
```diff
@@ -143,6 +154,14 @@ class BatchChangeValidations(
       case _ => Right(())
     }
 
+  def validateCreatorCancellation(
+      batchChange: BatchChange,
+      auth: AuthPrincipal): Either[BatchChangeErrorResponse, Unit] =
+    if (batchChange.userId == auth.userId) {
```
Comment (pauljamescleary) [inline modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala]: Should we allow super/support users to cancel?  I know they can reject which is effectively the same, so thinking out loud a little.

File: modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlBatchChangeRepository.scala [hunk 2]
```diff
@@ -90,7 +90,8 @@ class MySqlBatchChangeRepository
          |              bc.approval_status, bc.reviewer_id, bc.review_comment, bc.review_timestamp, bc.scheduled_time,
          |              SUM( case when sc.status like 'Failed' or sc.status like 'Rejected' then 1 else 0 end ) AS fail_count,
          |              SUM( case when sc.status like 'Pending' or sc.status like 'NeedsReview' then 1 else 0 end ) AS pending_count,
-         |              SUM( case sc.status when 'Complete' then 1 else 0 end ) AS complete_count
+         |              SUM( case sc.status when 'Complete' then 1 else 0 end ) AS complete_count,
+         |              SUM( case sc.status when 'Cancelled' then 1 else 0 end ) AS cancelled_count
```
Comment (mitruly) [inline modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlBatchChangeRepository.scala]: Are we opting for `cancelled` over `canceled`?

**Comment by mitruly** [comment]
Pretty good with this PR.

@pauljamescleary I had thought about the same thing before; it's more of a design approach and I think it makes sense to only allow the submitting user to cancel the changes.

File: modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlBatchChangeRepository.scala (hunk 2 above)
Comment (britneywright) [reply modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlBatchChangeRepository.scala]: cancelled 🙌🏽

File: modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala (hunk 1 above)
Comment (britneywright) [reply modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala]: I think only a user can cancel their own request. Super and Support users would reject. No need to mix the two.

Comment (rebstar6) [reply modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeValidations.scala]: +1, they can reject

File: modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeService.scala [hunk 3]
```diff
@@ -551,4 +561,16 @@ class BatchChangeService(
 
     batchChangeRepo.save(rejectedBatch).toBatchResult
   }
+
+  def cancelBatchChange(batchChange: BatchChange): BatchResult[BatchChange] = {
+    val cancelledSingleChanges = batchChange.changes.map(_.cancel)
+
+    // Update rejection attributes and single changes for batch change
+    val cancelledBatch = batchChange.copy(
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeService.scala]: do we care about any other metadata? must be this user, so dont need the user info, but what about cancelled time?

Comment (britneywright) [reply modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeService.scala]: Thought about cancelled time, I could go either way on it.

Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeService.scala]: Good point, I think we should save a time when it was cancelled.

Comment (rebstar6) [reply modules/api/src/main/scala/vinyldns/api/domain/batch/BatchChangeService.scala]: yeah, in general we cancel would only happen pretty immediately, but cant hurt to have. was wondering if we could leverage reviewTimestamp given that something cant be cancelled and any kind of other review, though the wording doesnt quite work

File: modules/api/functional_test/live_tests/batch/cancel_batch_change_test.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    assert_that(get_batch['changes'][0]['status'], is_('NeedsReview'))
+    assert_that(get_batch['changes'][0]['validationErrors'][0]['errorType'], is_('ZoneDiscoveryError'))
+
+    client.cancel_batch_change(result['id'], status=200)
+    get_batch = client.get_batch_change(result['id'])
+
+    assert_that(get_batch['status'], is_('Cancelled'))
+    assert_that(get_batch['approvalStatus'], is_('Cancelled'))
+    assert_that(get_batch['changes'][0]['status'], is_('Cancelled'))
+    assert_that(get_batch, has_key('cancelledTimestamp'))
```
Comment (mitruly) [inline modules/api/functional_test/live_tests/batch/cancel_batch_change_test.py]: Can we add the following assertion:
```
assert_that(get_batch, not(has_key('reviewTimestamp')))
```

Can we also do something similar for the existing reject batch change functional test?

**Pull request merged**